068-access-policies-in-jwt.md

```diff
@@ -0,0 +1,315 @@
+but this is not always the case.
+However, pretty much all identity providers support the concept of roles or claims.
+Roles are often used to define the access level of a user or service.
+For example, the database service is given the `admin` role, while the frontend service is given the `read` role. 
+In the application layer, the role is checked to see if the user has the necessary permissions when the user tries to perform an action.
+
+We can use fine-grained roles to represent ACLs.
+Instead of naming the role `admin`, which implicitly implies all permissions, we can name the role `*:topic:*:*` or if we want to limit the permissions - `*:topic:my_topic:read`.
+This way we are using the name of the role to encode the permissions the user has,
+instead of using a generic name such as `admin` and having the application layer understand what `admin` means.
+
+As an example we can define the roles for a service A in the IDP to be:
+```json
+"roles": [
+  "*:topic:input_topic:read",
+  "*:topic:output_topic:write",
+],
+```
+Now when these "roles" are passed in the JWT, we can use them as Kafka ACLs in order to authorize service A.
+
+### ACL Syntax
+
+A single ACL has the syntax `CLUSTER_NAME:RESOURCE_TYPE:RESOURCE_SPEC:PERMITTED_ACTIONS`.
+The separator `:` is chosen because it is a common separator in the Unix world.
+
+If the ACLs are passed as a string, they are separated by `,`.
+i.e., `CLUSTER_NAME:RESOURCE_TYPE:RESOURCE_SPEC:PERMITTED_ACTIONS,CLUSTER_NAME:RESOURCE_TYPE:RESOURCE_SPEC:PERMITTED_ACTIONS`
+
+- `RESOURCE_TYPE` can be `topic`, `group`, `transactional_id`, `delegation_token`, `user`, or the shortened versions `t`, `g`, `tid`, `dt`, `u`.
+- `PERMITTED_ACTIONS` is a list of a subset of `read`, `write`, `create`, `delete`, `alter`, `describe`, `cluster_action`, `describe_configs`, `alter_configs`, `idempotent_write`, `create_tokens`, `describe_tokens`, `all`, or the shortened versions `r`, `w`, `c`, `d`, `a`, `de`, `ca`, `dc`, `ac`, `iw`, `ct`, `dt`. The short version for `all` is `*`.
+The list items are separated by a `+`.
+- `CLUSTER_NAME` is the name of the cluster, and `RESOURCE_SPEC` is the name of the resource (topic/group).
+These fields can start or end with `*` to match any prefix/suffix.
+
+### Defaults
+If `CLUSTER_NAME` or `RESOURCE_SPEC` is empty, it is considered a wildcard `*`, which matches any cluster or resource.
+
+If `RESOURCE_TYPE` is empty, it takes the default value `topic`.
+
+If `PERMITTED_ACTIONS` is empty, no actions are allowed.
+
+The defaults for `CLUSTER_NAME`, `RESOURCE_SPEC`, and `RESOURCE_TYPE` are chosen for convenience.
+The default for `PERMITTED_ACTIONS` is none to prevent accidental access to all resources.
+
+### Examples of ACLs
+
+- `my_cluster:t:topic1:r+w` - allows reading and writing to `topic1` in the cluster `my_cluster`.
+- `:::` - denies all actions on all topics in all clusters.
+- `:::*` - allows all actions on all topics in all clusters.
```

Comment:
tbh I think the last two example are not readable. I think the empty defaults (for cluster and resource) or using `topic` if no resource type is specified don't work really well imho. I would use `*` or making the resource type mandatory.

Reply:
I somewhat agree. I think best practice would be to be explicit about all the fields. The only reasonable time to use such shorthand syntax would be when developing locally or debugging.